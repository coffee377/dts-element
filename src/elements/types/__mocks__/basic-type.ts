import {Type} from '../../type';

export class BasicType extends Type<any, any> {

  public get default_parameters(): any {
    return {};
  }

  public _emit(): string {
    return `[BasicType ${this.parameters.name}]`;
  }

}
