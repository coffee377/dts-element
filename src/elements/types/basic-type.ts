import {Stack} from '../../stack';
import {Type} from '../type';

export interface IBasicTypeRequiredParameters {
  name: string;
}

// tslint:disable-next-line no-empty-interface
export interface IBasicTypeOptionalParameters {}

export class BasicType extends Type<IBasicTypeRequiredParameters, IBasicTypeOptionalParameters> {

  public get default_parameters(): IBasicTypeOptionalParameters {
    return {};
  }

  public _emit(_stack: Stack): string {
    return this.parameters.name;
  }

}
