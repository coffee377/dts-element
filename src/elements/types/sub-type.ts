import {emit_brackets} from '../../helpers/emit-brackets';
import {Stack} from '../../stack';
import {Declaration} from '../declaration';
import {NamespaceDeclaration} from '../declarations/namespace-declaration';
import {Type} from '../type';

export interface ISubTypeRequiredParameters {
  target: Type;
}

export interface ISubTypeOptionalParameters {
  namespaces: NamespaceDeclaration[];
  path: (Declaration | Type)[];
}

export class SubType extends Type<ISubTypeRequiredParameters, ISubTypeOptionalParameters> {

  public get default_parameters(): ISubTypeOptionalParameters {
    return {
      namespaces: [],
      path: [],
    };
  }

  public _emit(stack: Stack): string {
    const {target} = this.parameters;
    const target_content = emit_brackets(target.emit(stack), target);
    const namespaces = this.parameters.namespaces
      .map((namespace: NamespaceDeclaration) => namespace.parameters.name);
    const path = this.parameters.path
      .map((element: Declaration | Type): string => `[${
        (element instanceof Declaration)
          ? JSON.stringify(element.parameters.name)
          : element.emit(stack)
      }]`)
      .join('');
    return `${[...namespaces, target_content].join('.')}${path}`;
  }

}
