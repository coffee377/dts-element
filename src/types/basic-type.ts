import * as ts from 'typescript';
import {IType} from '../collections';
import {ElementKind} from '../constants';
import {create_element, IElement} from '../element';
import {transform} from '../transform';

export interface IBasicTypeOptions {
  name: string;
  generics?: IType[];
}

export interface IBasicType
  extends IElement<ElementKind.BasicType>, IBasicTypeOptions {}

export const create_basic_type = (options: IBasicTypeOptions): IBasicType => ({
  ...create_element(ElementKind.BasicType),
  ...options,
});

// tslint:disable:ter-indent

export const transform_basic_type = (element: IBasicType, path: IElement<any>[]) =>
  ts.createTypeReferenceNode(
    /* typeName      */ element.name,
    /* typeArguments */ element.generics && element.generics.map(
                          generic => transform(generic, [...path, element]) as ts.TypeNode,
                        ),
  );
