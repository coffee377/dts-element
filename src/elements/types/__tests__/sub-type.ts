jest.unmock('../sub-type');

import {BasicType} from '../basic-type';
import {SubType} from '../sub-type';

describe('#emit()', () => {
  it('should return correctly', () => {
    expect(new SubType({target: new BasicType({name: 'A'})}).emit()).toMatchSnapshot();
  });
});
