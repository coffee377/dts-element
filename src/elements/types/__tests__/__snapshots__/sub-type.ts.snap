// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#emit() should return correctly 1`] = `"A.B.[emit-brackets [InterfaceType undefined]][\\"y\\"][[BasicType any]]"`;
